app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java

```diff
@@ -134,64 +156,107 @@ public void updateShouldRunDecision() {
+        boolean prefAlwaysRunInBackground = mPreferences.getBoolean(Constants.PREF_ALWAYS_RUN_IN_BACKGROUND, false);
+
+        // PREF_POWER_SOURCE
+        switch (prefPowerSource) {
+            case POWER_SOURCE_AC:
+                if (!isOnAcPower()) {
+                    Log.v(TAG, "decideShouldRun: POWER_SOURCE_AC && !isOnAcPower");
+                    return false;
+                }
+                break;
+            case POWER_SOURCE_BATTERY:
+                if (isOnAcPower()) {
+                    Log.v(TAG, "decideShouldRun: POWER_SOURCE_BATTERY && isOnAcPower");
+                    return false;
+                }
+                break;
+            case POWER_SOURCE_AC_BATTERY:
+            default:
+                break;
+        }
 
         // Power saving
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
             if (prefRespectPowerSaving && isPowerSaving()) {
+                Log.v(TAG, "decideShouldRun: prefRespectPowerSaving && isPowerSaving");
                 return false;
             }
         }
 
-        // Always run in background
-        if (!prefAlwaysRunInBackground) {
-            /**
-             * User did not specify run conditions in the options.
-             * The app is displaying a foreground activity and syncthing should run.
-             */
-            Log.v(TAG, "decideShouldRun: !prefAlwaysRunInBackground");
-            return true;
+        // Android global AutoSync setting.
+        if (prefRespectMasterSync && !ContentResolver.getMasterSyncAutomatically()) {
+            Log.v(TAG, "decideShouldRun: prefRespectMasterSync && !getMasterSyncAutomatically");
+            return false;
         }
 
-        // Run only when charging.
-        if (prefRunOnlyWhenCharging && !isCharging()) {
-            Log.v(TAG, "decideShouldRun: prefRunOnlyWhenCharging && !isCharging");
-            return false;
+        // Run on mobile data.
+        if (prefRunOnMobileData && isMobileDataConnection()) {
+            Log.v(TAG, "decideShouldRun: prefRunOnMobileData && isMobileDataConnection");
+            return true;
         }
 
-        // Run only on wifi.
-        if (prefRunOnlyOnWifi && !isWifiOrEthernetConnection()) {
-            // Not on wifi.
-            Log.v(TAG, "decideShouldRun: prefRunOnlyOnWifi && !isWifiOrEthernetConnection");
-            return false;
+        // Run on wifi.
+        if (prefRunOnWifi && isWifiOrEthernetConnection()) {
+            if (prefRunOnMeteredWifi) {
+                // Check if we are on metered wifi and if wifi whitelist run condition is met.
+                if (isMeteredNetworkConnection() && wifiWhitelistConditionMet(prefWifiWhitelistEnabled, whitelistedWifiSsids)) {
+                    Log.v(TAG, "decideShouldRun: prefRunOnWifi && isWifiOrEthernetConnection && prefRunOnMeteredWifi && isMeteredNetworkConnection && wifiWhitelistConditionMet");
+                    return true;
+                }
```

Comment:
You're right about this. I've corrected this.

Reply:
To be honest, this would be eaiser to read if it was:
```
if (runOnWifi && connectedToWifi) {
   if (!preferRunOnMetered && onMetered()) {
       return false;
   }
   if (wifiSsidWhitelisted) {
       return true
   }
}
```